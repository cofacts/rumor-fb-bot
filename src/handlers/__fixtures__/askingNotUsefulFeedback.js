export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
  },
};
