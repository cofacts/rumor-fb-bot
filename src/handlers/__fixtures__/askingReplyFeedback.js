export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};

export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const articleData = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};
