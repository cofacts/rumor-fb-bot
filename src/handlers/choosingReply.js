import gql from '../gql';
import {
  createPostbackAction,
  createReferenceWords,
  createTypeWords,
  ellipsis,
  getArticleURL,
} from './utils';
import ga from '../ga';

/**
 * The state that a user is choosing which reply to read
 * if she hasn't chosen one, we ask her to select the number of desired reply
 * if she has, we query the reply and render
 */
export default async function choosingReply(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;

  if (!data.foundReplyIds) {
    throw new Error('foundReplyIds not set in data');
  }

  const visitor = ga('FB-' + userId, data.selectedArticleText);
  visitor.screenview({ screenName: state });

  const selectedReplyId = data.foundReplyIds[event.input - 1];

  if (!selectedReplyId) {
    replies = [
      {
        type: 'text',
        content: {
          // Please enter 1 - {length} to choose a reply
          text: `請輸入 1～${data.foundReplyIds.length} 的數字，來選擇回應。`,
        },
      },
    ];

    state = 'CHOOSING_REPLY';
  } else {
    const {
      data: { GetReply },
    } = await gql`
      query($id: String!) {
        GetReply(id: $id) {
          type
          text
          reference
          createdAt
        }
      }
    `({ id: selectedReplyId });

    replies = [
      {
        type: 'text',
        content: {
          // Someone marked this message as "{type}" because
          text: `有人標記這個訊息 ${createTypeWords(GetReply.type)}，理由是：`,
        },
      },
      {
        type: 'text',
        content: {
          text: ellipsis(GetReply.text, 2000),
        },
      },
      {
        type: 'text',
        content: {
          text: ellipsis(createReferenceWords(GetReply), 2000),
        },
      },
      {
        type: 'text',
        content: {
          // These messages are provided by some nice volunteers. Please refer to
          // {articleURL} for more information, replies and references.
          text: `💁 以上訊息由好心人提供。建議至 ${getArticleURL(
            data.selectedArticleId
          )} 觀看完整的訊息內容、其他鄉親的回應，以及他們各自所提出的理由與出處。`,
        },
      },
      {
        type: 'buttons',
        content: {
          attachment: {
            type: 'template',
            payload: {
              template_type: 'button',
              // Is this reply helpful?
              text: '請問上面回應是否有幫助？',
              buttons: [
                // Yes
                createPostbackAction('是', 'y'),
                // No
                createPostbackAction('否', 'n'),
              ],
            },
          },
        },
      },
    ];
    // Track when user select a reply.
    visitor.event({ ec: 'Reply', ea: 'Selected', el: selectedReplyId });
    // Track which reply type reply to user.
    visitor.event({ ec: 'Reply', ea: 'Type', el: GetReply.type });

    data.selectedReplyId = selectedReplyId;
    state = 'ASKING_REPLY_FEEDBACK';
  }

  visitor.send();
  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
