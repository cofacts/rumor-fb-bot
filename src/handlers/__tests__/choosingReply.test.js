jest.mock('../../gql');

import choosingReply from '../choosingReply';
import * as apiResult from '../__fixtures__/choosingReply';
import gql from '../../gql';

it('should select reply by replyId', async () => {
  gql.__push(apiResult.oneReply);

  const params = {
    data: {
      searchedText: '貼圖',
      foundArticleIds: [
        'AWDZYXxAyCdS-nWhumlz',
        '5483323992880-rumor',
        'AV-Urc0jyCdS-nWhuity',
        'AVsh8u7StKp96s659Dgq',
      ],
      selectedArticleId: 'AWDZYXxAyCdS-nWhumlz',
      foundReplyIds: ['AWDZeeV0yCdS-nWhuml8'],
    },
    state: 'CHOOSING_REPLY',
    event: {
      type: 'postback',
      input: 1,
      timestamp: 1518964687709,
      postback: { data: '{"input":1,"issuedAt":1518964675191}' },
    },
    issuedAt: 1518964688672,
    userId: 'Uaddc74df8a3a176b901d9d648b0fc4fe',
    replies: [
      {
        type: 'text',
        text:
          '這篇訊息有：\n0 則回應認為其 ❌ 含有不實訊息\n0 則回應認為其 ⭕ 含有真實訊息\n0 則回應認為其 💬 含有個人意見\n1 則回應認為其 ⚠️️ 不在查證範圍\n',
      },
    ],
    isSkipUser: false,
  };

  expect(await choosingReply(params)).toMatchSnapshot();
});

it('should handle invalid reply ids', async () => {
  gql.__push(apiResult.oneReply);

  const params = {
    data: {
      searchedText: '貼圖',
      foundArticleIds: [
        'AWDZYXxAyCdS-nWhumlz',
        '5483323992880-rumor',
        'AV-Urc0jyCdS-nWhuity',
        'AVsh8u7StKp96s659Dgq',
      ],
      selectedArticleId: 'AWDZYXxAyCdS-nWhumlz',
      foundReplyIds: ['AWDZeeV0yCdS-nWhuml8'],
    },
    state: 'CHOOSING_REPLY',
    event: {
      type: 'text',
      input: '123',
      timestamp: 1518964687709,
    },
    issuedAt: 1518964688672,
    userId: 'Uaddc74df8a3a176b901d9d648b0fc4fe',
    replies: [
      {
        type: 'text',
        text:
          '這篇訊息有：\n0 則回應認為其 ❌ 含有不實訊息\n0 則回應認為其 ⭕ 含有真實訊息\n0 則回應認為其 💬 含有個人意見\n1 則回應認為其 ⚠️️ 不在查證範圍\n',
      },
    ],
    isSkipUser: false,
  };

  expect(await choosingReply(params)).toMatchSnapshot();
});

afterEach(() => {
  gql.__reset();
});
