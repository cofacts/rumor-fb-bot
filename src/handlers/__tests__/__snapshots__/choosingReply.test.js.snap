// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle invalid reply ids 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AWDZYXxAyCdS-nWhumlz",
      "5483323992880-rumor",
      "AV-Urc0jyCdS-nWhuity",
      "AVsh8u7StKp96s659Dgq",
    ],
    "foundReplyIds": Array [
      "AWDZeeV0yCdS-nWhuml8",
    ],
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
  },
  "event": Object {
    "input": "123",
    "timestamp": 1518964687709,
    "type": "text",
  },
  "isSkipUser": false,
  "issuedAt": 1518964688672,
  "replies": Array [
    Object {
      "content": Object {
        "text": "Please enter 1～1 to choose a reply.",
      },
      "type": "text",
    },
  ],
  "state": "CHOOSING_REPLY",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;

exports[`should select reply by replyId 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AWDZYXxAyCdS-nWhumlz",
      "5483323992880-rumor",
      "AV-Urc0jyCdS-nWhuity",
      "AVsh8u7StKp96s659Dgq",
    ],
    "foundReplyIds": Array [
      "AWDZeeV0yCdS-nWhuml8",
    ],
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": 1,
    "postback": Object {
      "data": "{\\"input\\":1,\\"issuedAt\\":1518964675191}",
    },
    "timestamp": 1518964687709,
    "type": "postback",
  },
  "isSkipUser": false,
  "issuedAt": 1518964688672,
  "replies": Array [
    Object {
      "content": Object {
        "text": "Someone marked this message as \\"⚠️️ OFF TOPIC\\" because:",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "􀂅 ⚠️️ This reply doesn't have any reference so it may not be credible. ⚠️️  􀂅",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "💁 These replies are provided by some nice volunteers. Please refer to https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz for more information, replies and references.",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "buttons": Array [
              Object {
                "payload": "y",
                "title": "Yes",
                "type": "postback",
              },
              Object {
                "payload": "n",
                "title": "No",
                "type": "postback",
              },
            ],
            "template_type": "button",
            "text": "Is this reply helpful?",
          },
          "type": "template",
        },
      },
      "type": "buttons",
    },
  ],
  "state": "ASKING_REPLY_FEEDBACK",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;

exports[`should throw error when reply ids are not set 1`] = `"foundReplyIds not set in data"`;
